### PR TITLE
Plugins to download from SFTP

### PR DESCRIPTION
**Pull request by ianstalk**
SFTP plugins patterned after ftp_list and ftp_download. This plugin requires the pysftp Python modules and its dependencies.

### CODE REVIEW

File: flexget/plugins/plugin_sftp.py [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+              get_size: True
+              files_only: False
+          dirs:
+              - '/path/to/list/'
+    """
+
+    schema = {
+        'type': 'object',
+        'properties': {
+            'config': {
```
Comment (gazpachoking) [inline flexget/plugins/plugin_sftp.py]: Do we need a separate key for this?

File: flexget/plugins/plugin_sftp.py [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+        url_prefix = 'sftp://%s%s%s/' % (login_str, host, port_str)
+        
+        log.debug('Connecting to %s' % host)
+
+        conn_conf = ConenctionConfig(host, port, username, password, private_key, private_key_pass)
+        
+        try:
+            sftp = sftp_connect(conn_conf)
+        except Exception as e:
+            log.error('Failed to connect to %s (%s)' % (host, e))
```
Comment (gazpachoking) [inline flexget/plugins/plugin_sftp.py]: Maybe this should raise a PluginError

File: flexget/plugins/plugin_sftp.py [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+            log.warn('Skipping unknown file %s' % path)
+
+    def on_task_download(self, task, config):
+        """
+        Task handler for sftp_download plugin
+        """
+        dependency_check()
+        config = self.prepare_config(config, task)
+
+        # Download entries by host so we can reuse the connection
```
Comment (gazpachoking) [inline flexget/plugins/plugin_sftp.py]: Should we be filtering to only try to process entries using the sftp uri scheme?

File: flexget/plugins/plugin_sftp.py (hunk 1 above)
Comment (ianstalk) [inline flexget/plugins/plugin_sftp.py]: I just did it that way to keep it similar to the existing FTP/FTPS plugin. Not sure why the author of that plugin did it that way, so I'm not particularly attached to the idea.

File: flexget/plugins/plugin_sftp.py [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+        host = parsed.hostname
+        username = parsed.username or None
+        password = parsed.password or None
+        port = parsed.port or 22
+
+        # get private key info if it exists
+        try:
+            private_key = entry['private_key']
+        except KeyError:
+            private_key = None
```
Comment (gazpachoking) [inline flexget/plugins/plugin_sftp.py]: Dicts already have a method for this, `private_key = entry.get('private_key')`

File: flexget/plugins/plugin_sftp.py [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+
+        try:
+            private_key = entry['private_key_pass']
```
Comment (gazpachoking) [inline flexget/plugins/plugin_sftp.py]: This meant to be `private_key_pass = ...`? Also, same note as above.

File: flexget/plugins/plugin_sftp.py [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+        except KeyError:
+            private_key_pass = None
+
+        if parsed.scheme == 'sftp':
+            config= ConnectionConfig(host, port, username, password, private_key, private_key_pass)
```
Comment (gazpachoking) [inline flexget/plugins/plugin_sftp.py]: Spaces around operators.

**Comment by gazpachoking** [comment]
Thanks!

**Pull request merged**